### PR TITLE
Overhaul for v7

### PR DESCRIPTION
**Pull request by hanjiexi**
Hi,

This aggregates all my suggested changes for v7 of the curriculum that we've been discussing in #387.

I know it's a lot! So don't feel bad about rejecting this pull request if it's not the direction you want the project to go in.

Cheers

### CODE REVIEW

**Comment by li-xinyang** [comment]
I like this new curriculum. It's updated and solid.

**Comment by ericdouglas** [comment]
I'll review your PR carefully, many thanks for your contribution!

About the _Change 'free' to 'little to no cost'_, it goes against one of our core values, that is to provide a free educational path.

It does **not** implicate that we will add bad courses only because they are free. If we don't have a good free option, we won't add any course for that matter until we have one that is both good (at least) **and** free.

**Comment by a1h05** [comment]
@ericdouglas 

> If we don't have a good free option, we won't add any course for that matter until we have one that is both good (at least) and free.

In cases in which free options does not exists but topic is important for gaining solid CS knowledge it would be great at least to mention topics( and probably corresponding courses ) somewhere at the end of page.

**Comment by TYRONEMICHAEL** [comment]
What do we do in the case where the paid alternative is a better course?

**Comment by real-gobs** [comment]
I think I agree with what @ericdouglas is saying, because some of us can't afford paid courses. So I think the best way is to present the two (both paid and free). So you choose which one best suits you.

**Comment by hanjiexi** [comment]
My approach was to treat quality and completeness of the curriculum as first-class requirements, with affordability as a second-class requirement. This would mean that the main track might have courses that cost money to access their materials, but free alternatives would be provided for those who couldn't afford them.

This leaves a grey area: courses whose *materials* are free but where some extra features, like graded tests/projects, were not. I reasoned that we can't regard these courses as truly costing money, because then we would have to say that some courses from MIT OCW cost money (you have to attend MIT if you want your tests graded!). Taking this a step further, free learning materials in general across the Internet don't necessarily provide the answers to everything, yet we still call them free.

So in evaluating the price of a course, I can only count access to the learning materials. (Note that for Pro CS, none of the above requirements matter, since it isn't part of the curriculum proper, it's just some suggestions of where to go next for professional development.)

The second thing I had to think about was whether *specifically* completing the courses listed in the curriculum was truly a hard requirement, or if it mattered more that the subject matters were learned one way or another. But everyone is starting from a different place. This resulted in my creating the [academic requirements](https://github.com/hanjiexi/computer-science/blob/v7.0.0/REQUIREMENTS.md) page.

That page lists the **hard requirements** of completing the curriculum, whereas following the courses should be thought of as a *soft* requirement, i.e. this is the preferred track for the majority of students, but there can be exceptions. Courses from [extras/courses.md](https://github.com/hanjiexi/computer-science/blob/v7.0.0/extras/courses.md) or readings from [extras/readings.md](https://github.com/hanjiexi/computer-science/blob/v7.0.0/extras/readings.md) can in some cases be a drop-in replacement.

So fulfilling the "hard requirements" just means learning those subjects, one way or another, free or paid. Fulfilling the "soft requirements" means following the courses, and here's how this ended up playing out:
- All primary coursework (Intro CS + all courses in Core CS + anything in Advanced CS minus capstone projects) can be completed for free
- The curriculum requires a project at the end of Core CS and at the end of Advanced CS. CS50 requires a final project too, so you can think of Intro CS as also requiring a project.
    - The Core CS project can be fulfilled for free (by creating your own) or by taking the capstone course (which you can pay for if you want your project graded)
    - The Advanced CS project should generally be fulfilled for free (by creating your own), but you could also take a paid capstone course under Advanced applications and, if the course's honor code allows it, share that project with the world as your final project.
    
TL;DR: As it stands right now, you can complete it for free, or you can pay to speed things up a bit.

**Comment by ericdouglas** [comment]
@hanjiexi by "free course" I mean "students must be able to access the resources for free".

**Comment by ericdouglas** [comment]
@hanjiexi I sent a dm on gitter to you. Let's discuss about this to make it happens ASAP! :tada:

**Comment by li-xinyang** [comment]
Wow! Finally merged! Thank you so much @hanjiexi & @ericdouglas  for your time and efforts.
